src/stream/mod.rs

```diff
@@ -471,3 +476,194 @@ impl Stream {
+        if buf.remaining() == 0 {
+            return Poll::Ready(Ok(()));
+        }
+
+        let fut = match self.read_fut.as_mut() {
+            Some(fut) => fut,
+            None => {
+                // read into a temporary buffer because `buf` has an unonymous lifetime, which can be
+                // shorter than the lifetime of `read_fut`.
+                let stream = self.stream.clone();
+                let mut temp_buf = Vec::with_capacity(buf.capacity());
+                self.read_fut.get_or_insert(Box::pin(async move {
+                    let res = stream.read(temp_buf.as_mut_slice()).await;
+                    match res {
+                        Ok(n) => { 
+                            temp_buf.truncate(n);
+                            Ok(temp_buf) 
+                        },
+                        Err(e) => Err(e),
+                    }
+                }))
+            }
+        };
+
+        loop {
+            match fut.as_mut().poll(cx) {
+                Poll::Pending => return Poll::Pending,
+                // retry immediately upon empty data or incomplete chunks
+                // since there's no way to setup a waker.
+                Poll::Ready(Err(Error::ErrTryAgain)) => {}
+                Poll::Ready(Err(e)) => return Poll::Ready(Err(e.into())),
+                Poll::Ready(Ok(read_buf)) => {
+                    let len = std::cmp::min(read_buf.len(), buf.remaining());
+                    buf.put_slice(&read_buf[..len]);
+                    self.read_fut = None;
+                    return Poll::Ready(Ok(()));
+                }
+            }
+        }
+    }
+}
+
+impl AsyncWrite for PollStream<'_> {
+    fn poll_write(
+        mut self: Pin<&mut Self>,
+        cx: &mut Context<'_>,
+        buf: &[u8],
+    ) -> Poll<io::Result<usize>> {
+        let fut = match self.write_fut.as_mut() {
+            Some(fut) => fut,
+            None => {
+                let stream = self.stream.clone();
+                let bytes = Bytes::copy_from_slice(buf);
+                self.write_fut.get_or_insert(Box::pin(async move {
+                    stream.write(&bytes).await
+                }))
+            }
+        };
+
+        match fut.as_mut().poll(cx) {
+            Poll::Pending => Poll::Pending,
+            Poll::Ready(Err(e)) => Poll::Ready(Err(e.into())),
+            Poll::Ready(Ok(n)) => {
+                self.write_fut = None;
+                Poll::Ready(Ok(n))
+            }
+        }
+    }
+
+    fn poll_flush(mut self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<io::Result<()>> {
+        match self.write_fut.as_mut() {
+            Some(fut) => match fut.as_mut().poll(cx) {
+                Poll::Pending => Poll::Pending,
+                Poll::Ready(Err(e)) => Poll::Ready(Err(e.into())),
+                Poll::Ready(Ok(_)) => {
+                    // XXX: is a data race between poll_write and poll_flush possible?
+                    self.write_fut = None;
```

Comment:
@tomaka since you've probably worked with AsyncWrite before, is a data race between poll_flush and poll_write a possibility? i.e. do I have to protect the field which is used by both functions using a Mutex? Thanks

Reply:
You don't need a `Mutex` in any case because both functions take `self` by `&mut`, however it is a bit subtle indeed.

`poll_flush` means "process the current queue of data"
`poll_write` means "queue as much of the given data as possible, and flush if it's not possible to queue more"